### DISCUSSION TITLE
Support for remote derive (or serde's from and into annotations)

### DISCUSSION THREAD

**Discussion by sunshowers** [diesel-rs/diesel#3919] [Ideas]
Hi there, and thanks for maintaining Diesel! I have a suggestion for a feature request that should hopefully not be too onerous.

In one of our main repositories, we define some common types in a `common` crate. This crate does not depend on Diesel (and we'd like to keep it that way -- these types are used in many places that don't need database access).

In our `db-model` crate, we would like to use these types directly in `derive(Selectable)` etc. But, as I'm sure you know, that runs into orphan rule issues -- it's not possible to implement a foreign trait for a foreign type.

The classic solution to that is to use a newtype. But now there are two very similar types floating around the system, and while there are ways to mitigate that it's still really annoying to manage.

While looking at ways to address this, I realized that serde may suggest a solution for this with its [remote derive](https://serde.rs/remote-derive.html) functionality. This could then be addressed with:

* Define a private newtype, say `RemoteType`, and derive `AsExpression` etc on it.
* Annotate `RemoteType` with `#[diesel(remote = "BaseType")]`.
* In the `Selectable` structs, use the `BaseType``, and annotate it with `#[serde(with = "RemoteType")]`.

Another alternative would be the [`from`, `try_from` and `into` annotations](https://serde.rs/container-attrs.html#from) would solve this problem.

* Define `RemoteType` and derive `AsExpression` etc on it.
* In the `Selectable` structs, use `BaseType`, and annotate it with `from` and `into` to the newtype.

Does either or both of them sound reasonable? I'm happy to try implementing it if so.

Apologies if this has been discussed already -- I searched through the tracker and couldn't find anything. Thanks!

**Comment by weiznich** [comment]
Thanks for filling this thread. I agree that this is an issue that deserves to be solved. Could you maybe try to provide an example of each suggested variant that shows:

* How the user facing code would look like
* What code would be generated by the derive

(That just needs to be a really simple example, and it does only need to cover a single derive)

I think that would make it much easier to understand what's actually suggested.
